### PR TITLE
Non-Standardized Supporting

### PR DESCRIPTION
**Pull request by AurevoirXavier**
I revamp my old implementation #211.

- All JSON tests were ignored under no-std. 
- `Reader`/`Tokenizer`/`serde`/`serde_json`should be enabled by `--features full-serde`.
  - I'm not sure if I should disable `Writer` in no-std, because I found it was used in `signature`.
- `Error::FromUtf8Error` removed, cause can't find any usage of it.
- All `HashMap` change to `BTreeMap` for no-std. **!!BREAKING change!!**

@sorpaas

### CODE REVIEW

File: ethabi/src/contract.rs [hunk 1]
```diff
@@ -6,33 +6,45 @@
...
 // except according to those terms.
 
-use crate::{errors, operation::Operation, Constructor, Error, Event, Function};
+#[cfg(not(feature = "std"))]
+use alloc::collections::{btree_map::Values, BTreeMap};
+#[cfg(feature = "full-serde")]
+use core::fmt;
+use core::iter::Flatten;
+#[cfg(feature = "std")]
+use std::collections::{btree_map::Values, BTreeMap};
```
Comment (sorpaas) [inline ethabi/src/contract.rs:15]: You technically don't need this line.

Add `extern crate alloc;` and the above `alloc` import will work on `std` as well.

**Comment by sorpaas** [comment]
I'd appreciate if you can add CI build run for no_std as well!

**Comment by vkgnosis** [comment]
Could you explain a bit on the motivation for no std support? What part of the crate is still useful without std? Why does HashMap have to change to BTreeMap?

**Comment by AurevoirXavier** [comment]
> Could you explain a bit on the motivation for no std support? What part of the crate is still useful without std? Why does HashMap have to change to BTreeMap?

I found many people use this in substrate runtime, so do I.

Some of them used it in EVM. And my team used it for the ethereum cross-chain bridge.

**Comment by AurevoirXavier** [comment]
> I'd appreciate if you can add CI build run for no_std as well!

Okay.

**Comment by sorpaas** [comment]
Please remove check and test command in CI, and add build.

**Comment by AurevoirXavier** [comment]
> Please remove check and test command in CI, and add build.

Please review.

**Comment by sorpaas** [comment]
CI is failing!

**Comment by AurevoirXavier** [comment]
> CI is failing!

I see. `cargo-hack` is pretty strict.

File: ethabi/Cargo.toml [hunk 2]
```diff
@@ -28,7 +28,6 @@ paste = "1"
 
 [features]
 default = [
-	"std",
```
Comment (sorpaas) [inline ethabi/Cargo.toml]: This is just wrong. By default `std` should be enabled.

Comment (AurevoirXavier) [reply ethabi/Cargo.toml]: > This is just wrong. By default `std` should be enabled.

But the `full-serde` contains std. Should I explicitly write it here again?

Comment (AurevoirXavier) [reply ethabi/Cargo.toml]: Okay. I found a corner case. You're right.

Comment (AurevoirXavier) [reply ethabi/Cargo.toml]: Sorry about ^^

The `full-serde` is enabled by `default`. So `std` is always enabled by `defualt`.

**Review by sorpaas — APPROVED** [review]

**Pull request merged**